xarray/plot/plot.py

```diff
@@ -429,6 +557,273 @@ def hist(
+    _is_facetgrid = kwargs.pop("_is_facetgrid", False)
+    if _is_facetgrid:
+        # Why do I need to pop these here?
+        kwargs.pop("y", None)
+        kwargs.pop("args", None)
+        kwargs.pop("add_labels", None)
+
+    _sizes = kwargs.pop("markersize", kwargs.pop("linewidth", None))
+    size_norm = kwargs.pop("size_norm", None)
+    size_mapping = kwargs.pop("size_mapping", None)  # set by facetgrid
+    cmap_params = kwargs.pop("cmap_params", None)
+
+    figsize = kwargs.pop("figsize", None)
+    subplot_kws = dict()
+    if z is not None and ax is None:
+        # TODO: Importing Axes3D is not necessary in matplotlib >= 3.2.
+        # Remove when minimum requirement of matplotlib is 3.2:
+        from mpl_toolkits.mplot3d import Axes3D  # type: ignore # noqa
+
+        subplot_kws.update(projection="3d")
+    ax = get_axis(figsize, size, aspect, ax, **subplot_kws)
+
+    _data = _infer_meta_data(darray, x, z, hue, hue_style, _sizes)
+
+    add_guide = kwargs.pop("add_guide", None)
+    if add_legend is not None:
+        pass
+    elif add_guide is None or add_guide is True:
+        add_legend = True if _data["hue_style"] == "discrete" else False
+    elif add_legend is None:
+        add_legend = False
+
+    if add_colorbar is not None:
+        pass
+    elif add_guide is None or add_guide is True:
+        add_colorbar = True if _data["hue_style"] == "continuous" else False
+    else:
+        add_colorbar = False
+
+    # need to infer size_mapping with full dataset
+    _data.update(
+        _infer_scatter_data(
+            darray,
+            x,
+            z,
+            hue,
+            _sizes,
+            size_norm,
+            size_mapping,
+            _MARKERSIZE_RANGE,
+        )
+    )
+
+    cmap_params_subset = {}
+    if _data["hue"] is not None:
+        kwargs.update(c=_data["hue"].values.ravel())
+        cmap_params, cbar_kwargs = _process_cmap_cbar_kwargs(
+            scatter, _data["hue"].values, **locals()
+        )
+
+        # subset that can be passed to scatter, hist2d
+        cmap_params_subset = {
+            vv: cmap_params[vv] for vv in ["vmin", "vmax", "norm", "cmap"]
+        }
+
+    if _data["size"] is not None:
+        kwargs.update(s=_data["size"].values.ravel())
+
+    # Plot the data. 3d plots has the z value in upward direction
+    # instead of y. To make jumping between 2d and 3d easy and intuitive
+    # switch the order so that z is shown in the depthwise direction:
+    axis_order = ["x", "z", "y"]
+    primitive = ax.scatter(
+        *[
+            _data[v].values.ravel()
+            for v in axis_order
+            if _data.get(v, None) is not None
+        ],
+        **cmap_params_subset,
+        **kwargs,
+    )
+
+    # Set x, y, z labels:
+    i = 0
+    set_label = [ax.set_xlabel, ax.set_ylabel, getattr(ax, "set_zlabel", None)]
+    for v in axis_order:
+        if _data.get(f"{v}label", None) is not None:
+            set_label[i](_data[f"{v}label"])
```

Comment:
how about something like 
```suggestion
            getattr(ax, "set_{axis}label")(_data[f"{axis}label"])
```

instead of incrementing `i`

Reply:
This is one of the unintuitive drawbacks I mentioned and why I started messing with `ax.view_init`. 
In 3d the z value corresponds to the `ax.set_ylabel`, because the y-axis has been rotated to go in the depthwise direction instead of vertical. But in 2d the y-axis is vertical and so corresponds to `ax.set_ylabel`.
And by conditionally incrementing `i` we handle both 2d and 3d. But it will feel a little weird if downstream users wants to change zlim or something.